### ISSUE TITLE
Inheritance and shared_ptr as container.

### ISSUE THREAD

**Issue by dtroitskiy** [kunitoki/LuaBridge3#51]
I was implementing constructors for classes with `shared_ptr` as container as described here https://kunitoki.github.io/LuaBridge3/Manual#343---shared_ptr-as-container
Problem is that I have two classes, one inherits from another, and this is where things go wrong, unless I misunderstand something.
I created separate test example to verify how things work, so the code may be as follows:

```
class Base : public std::enable_shared_from_this<Base>
{
public:
  Base(unsigned int id);
// ...
};

class Derived : public Base
{
public:
  Derived(unsigned int id, const std::string& name);
// ...
};
```

Now, when I create binding like this for `Base`:
```
getGlobalNamespace(L)
  .beginClass<Base>("Base")
  .addConstructorFrom<std::shared_ptr<Base>, void(unsigned int)>()
  .endClass();
```

everything's okay, but when I do

```
getGlobalNamespace(L)
  .deriveClass<Derived, Base>("Derived")
  .addConstructorFrom<shared_ptr<Derived>, void(unsigned int, const string&)>()
  .endClass();
```

I get this error:

`luabridge\detail\TypeTraits.h(63,5): error G56916779: static assertion failed due to requirement 'std::is_base_of_v<std::enable_shared_from_this<Derived>, Derived>'
      static_assert(std::is_base_of_v<std::enable_shared_from_this<T>, T>);`

First thing I tried is to make multiple inheritance for `Derived`, i.e.
`class Derived : public Base, public std::enable_shared_from_this<Derived>`.

Then I get this error

`luabridge\detail\TypeTraits.h(69,19): error G417914D0: member 'shared_from_this' found in multiple base classes of different types
          return t->shared_from_this();`

I google about `std::enable_shared_from_this` and inheritance in general, this seems to be good discussion https://stackoverflow.com/questions/657155/how-to-enable-shared-from-this-of-both-parent-and-derived
So there I found that simplest solution could be to override `shared_from_this()` in `Derived` like this:

```
shared_ptr<Derived> Derived::shared_from_this()
{
  return static_pointer_cast<Derived>(Base::shared_from_this());
}
```

but when I do it like this, I get `std::bad_weak_ptr` thrown in runtime when I try to pass instance of `Derived` into C++ function bound to Lua, more specifically, right when `Base::shared_from_this()` is called.

After some experimentation, turned that that I don't need to inherit `public std::enable_shared_from_this<Derived>`, but then I have to comment `std::is_base_of_v` check at `TypeTraits.h(63)`. And then it works.

Full source of my example is here https://www.dropbox.com/s/qplxkfmmtpbhyla/SharedFromThisTest.zip?dl=0
With `TypeTraits.h(63)` line commented it builds and runs no problem. As soon as you uncomment the line, you get the `static_assert` error I posted above.
I'm not sure whether I'm doing everything correct and as intended in this example. But, if compared to `RefCountedObjectPtr`, it works no problem out of the box with only `Base` inheriting from it. So I assume something similar should work for `shared_ptr` as well.

**Comment by kunitoki** [comment]
Yes indeed, there might be something fishy in that static assert, in the case of enable_shared_from_this, the check fails. Will look into it,thanks for reporting.

**Comment by kunitoki** [comment]
This PR should fix it https://github.com/kunitoki/LuaBridge3/pull/52

**Comment by dtroitskiy** [comment]
Yep, works for me, thanks for quick fix!

Small note though, I didn't run tests, but just from quick look at PR's code, I see that you test B like this
```
EXPECT_TRUE(runLua("result = test.B(2)"));
auto a4 = result<std::shared_ptr<A>>();
```
That doesn't seem 100% correct to me, because looks like you get `shared_ptr<A>` and not `shared_ptr<B>`. For it to be tested correctly, I assume you would need some unique member in `B` that `A` doesn't have and then get `shared_ptr<B>` and access that specific member of B. But that would also require from you to do what I did in my example: overriding (or probably more correct term is _hiding_ / _shadowing_, since it's not virtual, not sure) `shared_from_this()` method. So it would be something like
```
class B : public A
{
public:
    B() = default;

    B(int newX)
        : A(newX)
    {
    }
    
    int uniqueMethodOfB()
    {
        return 42;
    }
    
    std::shared_ptr<B> B::shared_from_this()
    {
        return std::static_pointer_cast<B>(A::shared_from_this());
    }
};
```
And then test calling `B::uniqueMethodOfB()` or do however you like it.

But if I think about it, that shouldn't make any difference for the actual fix. Looks like burden of overriding `shared_from_this()` should be on the shoulders of class developer anyway. Unless you want to specifically statically check in `TypeTraits.h` that this specific class being bound has its own `shared_from_this()` method that returns exactly this class pointer.

**Comment by kunitoki** [comment]
Yes thanks, makes sense. I've added something that should mitigate the need of having to override shared_from_this

**Comment by dtroitskiy** [comment]
I see. Well, I don't know what to recommend here, probably calling panic handler is correct thing to do, and it's up to LuaBridge user to watch with which types he works to avoid it. If I understand correctly, normally it should work fine when there is inheritance between classes, either polymorphic or not.

Speaking of inheritance, I noticed one more thing yesterday, looks like `static_assert` doesn't work anymore with classes without it, can you please check that? Previously in my code I used only two classes with `std::enabled_shared_from_this`, where B inherited from A, and yesterday I added one more class without any inheritance, well, except inheritance from `std::enabled_shared_from_this`, of course, and looks like `is_base_of_template_v` didn't work well on it.

**Comment by kunitoki** [comment]
Do you have an example that fails ?

**Comment by dtroitskiy** [comment]
Hm, weird, can't reproduce it anymore. Yesterday, when I had some error about it, I commented that `static_assert` line again, and it worked. And now I uncommented it, I still have these same classes without inheritance and code compiles without errors. So there's a chance that was false alarm or I messed up something in my code.

I also verified that bindings work correctly without `shared_from_this()` override, so I think this issue may be closed now.
Thanks again!

**Comment by dtroitskiy** [comment]
Sorry, I will need to reopen this issue, although thing I encountered doesn't have to do anything with inheritance, but it's still related to using `shared_ptr` as container, so I guess it's better to post it here rather than creating new issue.
And that's not critical, because there's easy workaround, but if it can be fixed, it would be nice.

So, something's wrong with binding properties of a class which are `shared_ptr`s, and this problem appears only when you bind property as plain field, and not when as property with getter method.

Minimum code example should be this:
```
// Test.h
class Test
{
public:
  class Nested : public std::enable_shared_from_this<Nested>
  {
  public:
    inline int getValue() const { return 42; }
  };

  Test();

  inline std::shared_ptr<Nested> getSpNested() const { return spNested; }

  static void bindToLua();

private:
  Nested nested;
  std::shared_ptr<Nested> spNested;
};

// Test.cpp
#include "luaState.h"
#include "luabridge/LuaBridge.h"
#include "Test.h"

using namespace luabridge;

Test::Test()
{
  spNested = std::make_shared<Nested>();
}

void Test::bindToLua()
{
  const auto L = getLuaState();

  if (!getGlobal(L, "Test").isTable())
  {
    getGlobalNamespace(L)
      .beginClass<Test>("Test")
        .addConstructor<void()>()
        .addProperty("nestedAsField", &Test::nested)
        .addProperty("spNestedAsField", &Test::spNested)
        .addProperty("spNestedAsFuncProp", &Test::getSpNested)
      .endClass()
      .beginClass<Test::Nested>("TestNested")
        .addFunction("getValue", &Test::Nested::getValue)
      .endClass();
  }
}
```
By the way, using nested class here is unrelated, it can be any class, here it's nested just for simplicity.

And then what's interesting, if you try to access `test.nestedAsField` in Lua, it works no problem, because it's simple class instance, not `shared_ptr`, but when you access `test.spNestedAsField`, which is `shared_ptr`, it fails without any error description. At the same time binding this property as getter method, i.e. `test.spNestedAsFuncProp`, works no problem.
So in Lua it looks like:
```
local test = Test()
print(test.nestedAsField:getValue()) -- Good.
print(test.spNestedAsFuncProp:getValue()) -- Good.
print(test.spNestedAsField:getValue()) -- Fails here.
```

Just in case, I updated my previous example code with that new `Test` class - https://www.dropbox.com/s/qplxkfmmtpbhyla/SharedFromThisTest.zip?dl=0.

**Comment by kunitoki** [comment]
Fixed in #57
